File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool isBalanced = true;
        dfs(root, isBalanced);
        return isBalanced;
    }

    int dfs(TreeNode* node, bool &isBalanced)
    {
        if (!node) return 0;

        int left = dfs(node->left, isBalanced);
        int right = dfs(node->right, isBalanced);

        if (abs(left - right) > 1) 
            isBalanced = false;

        return 1 + max(left, right);
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {

    public bool IsBalanced(TreeNode root) {
        bool isBalanced = true;
        DFS(root, ref isBalanced);
        return isBalanced;
    }
    public int DFS(TreeNode root, ref bool isBalanced)
    {
        if (root == null) return 0;

        int l = DFS(root.left, ref isBalanced);
        int r = DFS(root.right, ref isBalanced);

        if (Math.Abs(l - r) > 1)
        {
            isBalanced = false;
        }
        return 1 + Math.Max(l, r);
    }
}
